Add PSA hash driver dispatch specific testing

ronald-cron-arm opened the pull request:
## Description
As other cryptographic operations for which driver delegation is yet supported, add some specific testing of the dispatch for hash operations in test_suite_psa_crypto_driver_wrappers.

The PR moves the calls to the test driver entry points that actually do the job (mbedtls_transparent_test_driver_hash_xyz functions) from the driver wrapper (part of PSA core) to the test file hash.c. This will ease the move to a Mbed TLS test driver compiled separately.

## Status
READY

## Requires Backporting
NO, PSA only.

## Steps to test or reproduce
CI only,

stevew817 reviewed (approved):
LGTM, and I can see the reason for this in Mbed TLS context. I do however wonder how this will end up being integrated into on-device testing. E.g. when these cases get run when another accelerator is compiled in, the 'no fallback' tests will fail, since there's another accelerator present to pick up the job when the test driver fails (or the other accelerator gets picked before the test driver, in which case not even the hit counter will get updated).

gilles-peskine-arm commented on tests/include/test/drivers/hash.h:
```diff
@@ -0,0 +1,79 @@
+
+extern test_driver_hash_hooks_t test_driver_hash_hooks;
+
+psa_status_t test_transparent_hash_compute(
```
Even for testing only, I think we should avoid non-namespaced identifiers. They could conflict with the test framework of an embedded platform, for example. What't the motivation for not having a `mbedtls_` prefix?

I'd be ok with non-namespaced macros and types if they're internal to the library or tests, since they could only conflict with a header that our code includes. But names with external linkage can conflict with everything else that goes onto the platform.

ronald-cron-arm replied:
There is no special motivation for not having the `mbedtls_` prefix. I've just followed what we've been doing so far in the test drivers (see cipher.h, signature.h ...). I am going to add the prefix to all of them as I agree it will be better.

ronald-cron-arm replied:
I've addressed this now.

gilles-peskine-arm reviewed (changes requested):
LGTM except for the namespacing

ronald-cron-arm commented:
Following the merge of the PR on PSA AEAD (#4257) this PR had conflicts with `development`. I saved the version before rebase [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-hash-dispatch-testing-1). The rebase was simple, just conflicts due to new code added "at the same place" in some files in this PR and in #4257. After rebasing I also reworked the commit "tests: psa: Add mbedtls/MBEDTLS prefix to test driver symbols" to do the changes for AEAD test driver code as well.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.data:
```diff
@@ -243,3 +243,43 @@ aead_decrypt:PSA_KEY_TYPE_AES:"a0ec7b0052541d9e9c091fb7fc481409":PSA_ALG_GCM:"00
+
+Hash multi-part: SHA-256, INSUFFICIENT_MEMORY
+depends_on:MBEDTLS_PSA_BUILTIN_ALG_SHA_256
+hash_compute:PSA_ALG_SHA_256:"b0bd69":"4096804221093ddccfbf46831490ea63e9e99414858f8d75ff7f642c7ca61803":PSA_ERROR_INSUFFICIENT_MEMORY::PSA_ERROR_INSUFFICIENT_MEMORY
```
Please change `::` to `:` here an in one other test case. The fact that it works is an undocumented quirk which, digging into it, is only possible because `escaped_split` in `generate_test_code.py` is somewhat buggy.

While you're at is please fix the preexisting `::` in `test_suite_ssl.data`.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -936,3 +936,201 @@ exit:
+        ASSERT_COMPARE( output, output_length, hash->x, hash->len );
+    }
+
+    /*
```
Having all of these cases in the same function is fragile: sooner or later I expect we'll see cases where the core wants to change the error returned by the driver, in a way that depends on the stage of the operation. So we may want to have different expected statuses for a given forced status for e.g. setup vs update. The common preparation is small, so there's little gain in having a single test function covering everything.

Not a blocker, we can improve this later.

gilles-peskine-arm commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -37,7 +37,7 @@
 #define PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT
 #endif
-#include "test/drivers/test_driver.h"
+#include "test/drivers/mbedtls_test_driver.h"
```
Cosmetic: I don't think it made sense to rename this header file.

gilles-peskine-arm reviewed (changes requested):
Please get rid of the `::` in test data before they cause some mysterious error. Other than that LGTM.

ronald-cron-arm commented:
I've addressed Gilles' comments with additional commits and the result can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-hash-dispatch-testing-2). Then I've rebased on top of development and resolved the conflicts between development and the commit doing some significant renaming: "tests: psa: Add mbedtls/MBEDTLS prefix to test driver symbols".

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -1151,6 +1151,62 @@ exit:
+    PSA_ASSERT( psa_crypto_init( ) );
+
+    /*
+     * Update none active operation, the driver shouldn't be called.
```
Cosmetic: “none active” → “inactive” (here and in 2 other places)

ronald-cron-arm replied:
I've just done the change.

stevew817 reviewed (approved):
Still looks good. Assuming it'll create a merge conflict for #4247.
